### ISSUE TITLE
Turing 

### ISSUE THREAD

**Issue by Farnazmdi** [meyer-lab/DrugResponseModel.jl#50]

**Comment by Farnazmdi** [comment]
I first estimated the parameters as usual in DDEmodel, and then for one trial (here trial 6) I did the turing and used truncated normal distribution within the range of some multiples of the estimated parameter, this is the result.

 
![image](https://user-images.githubusercontent.com/34812750/66091453-b04ac780-e53c-11e9-9cac-030dc339e556.png)

this is a part of the code:
![image](https://user-images.githubusercontent.com/34812750/66091485-d53f3a80-e53c-11e9-88a6-cd593605fd16.png)

I am not sure how to interpret it, I compared it with another example I got [here](https://github.com/saumyagshah/JupyterNBTuringExamples/blob/master/time_series_ARIMA.ipynb)

**Comment by Farnazmdi** [comment]
This is in this [notebook](https://github.com/meyer-lab/DrugResponseModel.jl/blob/turing/src/DDEmodel.ipynb) @aarmey Could you please take a look at it and let me know if anything doesn't make sense? I can't figure out what part is wrong, but if you look at the last cell, there shouldn't be NaNs and Inf I guess.

**Comment by Farnazmdi** [comment]
![image](https://user-images.githubusercontent.com/34812750/66169392-15fa8a80-e5f5-11e9-9060-8b29be354a0a.png)

![image](https://user-images.githubusercontent.com/34812750/66169431-31fe2c00-e5f5-11e9-80ad-e59dac89dad7.png)

**Comment by Farnazmdi** [comment]
@aarmey  It seems working!

**Comment by aarmey** [comment]
Looks promising! My guess is that sigma is the error of the fit (think DiffEqBayes is assuming data is normally distributed...)

**Comment by Farnazmdi** [comment]
Dr. Meyer, I updated the packages, and it seems in the new version there is some problem. It errors and seems something internal is breaking. I ran an example from their repository in their tests, it also breaks. Could you please take a look at it whenever convenient? [Notebook](https://github.com/meyer-lab/DrugResponseModel.jl/blob/turing/src/DDEmodel.ipynb)

**Comment by aarmey** [comment]
From Turing or DiffEqBayes? What's happening isn't clear to me. If the tests for either package fail I'd bring it up as an issue. Can you just downgrade?

**Comment by Farnazmdi** [comment]
I guess it's Turing. I have a few examples I picked up from a video of their tutorial, and they are not DEproblems, they just define their model and do the inference with Turing. This even doesn't work and the problem has something to do with `sampleable model` I think they have changed some setting of the arguments. This is the error I get for the examples and my own DEproblem:
`MethodError: no method matching sample(::Turing.Model{Tuple{:omega,:alpha,:beta,:h},Tuple{:r,:TV},getfield(Main, Symbol("###inner_function#502#42")),NamedTuple{(:r, :TV),Tuple{Array{Float64,1},Type{Array{Float64,1}}}},NamedTuple{(:r,),Tuple{Symbol}}}, ::NUTS{Turing.Core.ForwardDiffAD{40},(),AdvancedHMC.Adaptation.DiagEuclideanMetric})
Closest candidates are:
  sample(::ModelType<:Sampleable, ::AlgType<:Turing.Inference.InferenceAlgorithm, !Matched::Integer; resume_from, kwargs...) where {ModelType<:Sampleable, AlgType<:Turing.Inference.InferenceAlgorithm} at /home/farnazm/.julia/packages/Turing/fbY6B/src/inference/Inference.jl:147
  sample(!Matched::MCMCChains.AbstractChains, ::Any; replace, ordered) at /home/farnazm/.julia/packages/MCMCChains/FusD4/src/sampling.jl:18
  sample(::ModelType<:Sampleable, !Matched::SamplerType<:Turing.Interface.AbstractSampler, !Matched::Integer; kwargs...) where {ModelType<:Sampleable, SamplerType<:Turing.Interface.AbstractSampler} at /home/farnazm/.julia/packages/Turing/fbY6B/src/interface/Interface.jl:170
  ...

Stacktrace:
 [1] top-level scope at In[5]:26`. 
I think there is something in their update that is breaking, but it is also possible that I have made a mistake and I don't know.

**Comment by aarmey** [comment]
Can you post the example model that doesn't run? That's useful for figuring out the problem because it's presumably simpler.

If you open an issue, it should come with a simple example of code that should work but doesn't.

**Comment by Farnazmdi** [comment]
Will do right away.

**Comment by Farnazmdi** [comment]
Ok, they made a PR and their master in `DiffEqBayes` is working, so until they release their new version, I cloned their master and use it. 
The thing with using turing for Hill is that, I need to have _one_ DEProblem to use the turing interface, but for the Hill, we have 8 separate DEProblems each for each concentration and we are collecting the cost for each in a for loop, so I need to find a way to put them all together and have _one_ problem to pass it to the `turing_interface`.

**Comment by aarmey** [comment]
For Hill, you'd need to directly work with Turing.jl. `turing_inference` just sets up a Turing.jl program, but that can be tricky: https://github.com/JuliaDiffEq/DiffEqBayes.jl/blob/192ab9613c0de534f6740df1d7bcfea23e40e6e8/src/turing_inference.jl

I think just using it for the DDE model is most helpful anyway, since the DDE model is underdetermined (too many parameters), while our problem with the Hill model is it is overdetermined (not flexible enough).

**Comment by Farnazmdi** [comment]
Yes exactly, that requires a different type of model and might take a lot of time; I will spend some time on it for DDE for all of the drugs. For DDE, actually each time we optimize for 6 parameters, compared to Hill it has fewer parameters; since each time for Hill, we are optimizing for 12 parameters. You are right, I should do it just for DDE. I will make the pull request as soon as possible.

**Comment by Farnazmdi** [comment]
Dr. Meyer, is this the only part remaining? Do you think we should add any other thing to the model or the figures? @aarmey

**Comment by aarmey** [comment]
At the moment I think this covers remaining items. I also don't think this will change our conclusions, so you're time is probably better spent on the lineage work until we can sit down with the Heiser lab.

**Comment by Farnazmdi** [comment]
Sure, thank you. I am also willing to do the experiments to get the data for the lineage project. (I am guessing it is easy to do for me as a beginner)

**Comment by Farnazmdi** [comment]
It looks like Turing is not possible with the new representation of the model (using ArrayOperators). But otherwise for the 2-equation model works fine.
![image](https://user-images.githubusercontent.com/34812750/70835142-0942a180-1db1-11ea-94c0-70057b9da160.png)
Should I return the model representation to the old version @aarmey ?

**Comment by aarmey** [comment]
How is it not possible?

**Comment by aarmey** [comment]
What is the error?

**Comment by Farnazmdi** [comment]
Starts with this and it is very long.
![image](https://user-images.githubusercontent.com/34812750/70835517-317ed000-1db2-11ea-9494-a856958a04b0.png)

**Comment by Farnazmdi** [comment]
It is in [here](https://github.com/meyer-lab/DrugResponseModel.jl/blob/turingODE/ODEmodel.ipynb)

**Comment by aarmey** [comment]
Ok, this looks like just a type error—should be easy to fix.

Please make a test that runs this fitting on a new branch, and I'll fix it so that it runs.

**Comment by Farnazmdi** [comment]
Thank you, will do right away

**Comment by aarmey** [comment]
Ah... hit a limitation in DifferentialEquations.jl... the `DiffEqArrayOperator` interface uses a library that relies on LAPACK matrix math, and so can't use Dual numbers. So, we'll have to go back to the old ODE structure like you did.

**Comment by aarmey** [comment]
I'd recommend using Turing directly, rather than through DiffEqBayes. You'll have to do this anyway for the 4ODE model, since the model-data comparison doesn't match the expected structure. DiffEqBayes also loads STAN and other Bayesian packages, so it's huuuuuge.

**Comment by Farnazmdi** [comment]
I was going to use the solutions as the data for the 4-equation model. But sure, I will switch to Turing.
Thanks
